tsdb/importer/import.go

```diff
@@ -0,0 +1,310 @@
+			continue
+		case textparse.EntryComment:
+			continue
+		default:
+		}
+		_, currentTimestampMicroS, val := parser.Series()
+
+		// The text parser converts all timestamps to microseconds.
+		// TSDB looks for timestamps in milliseconds.
+		var currentTimestampMs int64
+		if currentTimestampMicroS == nil {
+			currentTimestampMs = 0
+		} else {
+			currentTimestampMs = *currentTimestampMicroS / 1e3
+		}
+
+		var lset prom_labels.Labels
+		_ = parser.Metric(&lset)
+
+		sample := &tsdb.MetricSample{TimestampMs: currentTimestampMs, Value: val, Labels: tsdb_labels.FromMap(lset.Map())}
+
+		// Run this flow when current sample is outside the DB time limits
+		// OR
+		// when there are no blocks in the DB.
+		if sample.TimestampMs < dbMint || sample.TimestampMs > dbMaxt || len(blockTimes) == 0 || blockTimes == nil {
+			// In the case where we are importing data from before the DB
+			// start time, we could run into a case where the imported data
+			// could overlap with current data. This flag sets up the proper
+			// flow for that case.
+			if sample.TimestampMs < dbMint {
+				haveSamplesBeforeDbStart = true
+			}
+			// Take care of samples that are on either side of the DB's time limits.
+			if len(currentBlock) > 0 {
+				startTime, endTime := currentBlock[0].TimestampMs, currentBlock[len(currentBlock)-1].TimestampMs
+				// Check if adding current sample will push current block over 2h, if so, cut and push
+				// OR
+				// If we already have samples in the case where the current sample just goes past the
+				// DB maxt - then we must push the current blocks to disk, before accepting the new one.
+				if (sample.TimestampMs-startTime) >= tsdb.DefaultBlockDuration ||
+					haveSamplesInsideBlockTimeRange {
+					var parts [][]*tsdb.MetricSample
+					if haveSamplesInsideBlockTimeRange {
+						parts = partitionBlocks(currentBlock, endTime-startTime)
+					} else {
+						parts = partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+					}
+					for _, part := range parts {
+						path, err := tsdb.CreateBlock(part, dbPath, logger)
+						if err != nil {
+							return blockPaths, err
+						}
+						blockPaths = append(blockPaths, path)
+					}
+					currentBlock = []*tsdb.MetricSample{}
+					if haveSamplesInsideBlockTimeRange {
+						haveSamplesInsideBlockTimeRange = false
+					}
+				}
+			}
+			currentBlock = append(currentBlock, sample)
+			continue
+		}
+		// This will only be triggered in the case where we have some
+		// samples from before the DB mint, that also overlap with the
+		// DB blocks times.
+		if haveSamplesBeforeDbStart {
+			parts := partitionBlocks(currentBlock, tsdb.DefaultBlockDuration)
+			for _, part := range parts {
+				path, err := tsdb.CreateBlock(part, dbPath, logger)
+				if err != nil {
+					return blockPaths, err
+				}
+				blockPaths = append(blockPaths, path)
+			}
+			currentBlock = []*tsdb.MetricSample{}
+			// Once this is set to false in here, it will never be triggered again,
```

Comment:
I don't see how this is the case, it can happen again on the next sample.

Reply:
The flag is set to true only when it encounters entries before the DB mint, and since we deal with all such samples before we reach this block, the flag will not be set to true again once we get to this point.

Reply:
I don't see how that's the case, the input is not going to be ordered by time.

Reply:
Based on the approach that was suggested, I currently 'stream' the input and flush to disk as soon as I fill up a block. 

If inputs are not sorted by time, then this approach fails.

But I think that the input data should be sorted when fed to the importer, else we would have to go back to the old in-memory approach, and then sort the data to properly align the incoming blocks.

Reply:
You're using OpenMetrics, data will not be sorted in time overall

Reply:
I quickly tested switching over to the Prometheus parser to see if it would enforce a time series order, and that does not do it either. Open Metrics at least complains about series' being out of bounds. Do you have any suggestions on how to tackle this then?

One way I can think is to keep track of all existing blocks' ranges (if they exist), and then append samples to the appropriate block as we encounter them, but that runs into the issue of keeping all samples in memory until we have read the entire input data.

Reply:
Another approach I can think of would build off keeping track of existing blocks' ranges - basically, we could have a cutoff number of samples after which we flush them to disk (as new blocks aligned with the existing blocks), and then repeat the process until we have read all the input data. Any samples encountered after each flush to disk, that align with the newly created blocks, would be appended to the same newly created block instead of creating a new one each time.

Reply:
It'd have to be two pass. One to split things into blocks of time, the second to create the blocks.

Reply:
Makes sense, I will rework this PR.